### DISCUSSION TITLE
Slow download speed

### DISCUSSION THREAD

**Discussion by Anchovy8629** [9001/copyparty#29] [Q&A]
Hello, i'm not a programmer.. I'm hosting with my windows 10 PC using python and i have 900mb in upload. My friends have 900mb in download.. no one is using wifi.. I don't know why the download speed is so slow around 10mb.. Any tips? Thanks.

**Comment by 9001** [comment]
Hey! There's a lot of reasons why this could be the case, so let's try to go through some of them.

1. There's a big difference between megabit and megabyte! One megabyte is the same as 8 megabit. We often use megabit when talking about network speeds (because bigger numbers look better), but we use megabyte when talking about speeds of actual downloads / uploads. So I bet that your network speed is measured in megabit, which means that the highest speed you can possibly get is actually 110 megabyte per second.

But 110 is still way more than the 10 that you mentioned, so let's keep going...

2. The next likely bottleneck is the wifi you mentioned. Even if your own wifi is not busy at all, it is possible that people using other wifi networks nearby (for example your neighbors) can slow down your wifi.

Technically, this depends on which "channel" your wifi is on, and how busy that channel is (how many other wifi networks nearby are using the same channel) -- and the amount of channels to choose from is fairly limited.

So please try using a cable instead -- both you and your friend at the same time.

3. Now we can start suspecting your ISP, and the quality of the network between you guys... It is possible that the internet route between your computers is under high load. Like wifi, this is something that can change over the day -- so if you for example notice the speed gets better late at night, then this could be it.

One thing you could try is to have your friend download just one single big file from you (and nothing else at the same time), and make a note of how fast it was -- for example 10, like you mentioned. And then afterwards, try downloading two big files at the same time. We would expect that both of those files will download at around half the previous speed, so 5 for each file -- the total speed should not be higher. But, if you are getting higher speeds than that, for example 7 and 8 (so 15 total), then that could indicate that the bottleneck is the ISP / route.

Let us know how it goes!

**Reply by Anchovy8629** [reply]
Thanks for the reply, wifi is not a factor we are all using cable. If i test my connection on speedtest.net i get 935Mbps on both up and down. We all live in the same city. 

I did the test you suggested, with multiple files downloading i get more speed.. i just tried again and a single file averages at 4mbs, if i add others they will average the same. When i first start them i see an high spike of bandwidth for a few seconds, (20-25mbs) then goes to average 4mbs. 

 If the ISP is limiting my upload speed then why do i get full bandwidth on the speedtest?
On google drive for example uploading a single file it can go at around 50mbs (which is still low compared to 120mbs) but still better than what im getting with P2P 

There is anything i can do? more tests? Or do i just contact my ISP? i don't even know how to explain the problem to them.. Thanks.

**Comment by 9001** [comment]
Cool that we figured it out :>

 this is a fairly common problem, especially over long distances, and/or especially when you are connecting to servers which are running on budget plans. So I guess home connections kinda qualify for that description :p

Unfortunately I don't think you can do a lot about the actual problem... the ISP will probably say that you should pay for a business/enterprise connection or something, if you need good peering.

But! the good news is that you can usually get around it by sending data over several tcp connections instead, like we did with the multiple parallel downloads. This means that protocols like BitTorrent will be super fast. And although copyparty speaks HTTP (like most/all websites do), it has a somewhat unique upload feature which should get around the problem very well.

EDIT: Speedtest and Google Drive are much faster because they pay for premium connections, and they have a lot of servers so they're sending you to the one that's the closest to you -- and it's possible that speedtest also does the trick of using multiple connections, but i'm not sure about that part.

I suspect that if someone uploads files to your copyparty, they should get really good speeds, hopefully/probably more than Google Drive even -- would be great if you could verify that! The upload speed should go way up if they increase the "parallel uploads" to around 8 or maybe a bit more.

Downloading from copyparty, however, will only use one connection for each file. This is not a copyparty limitation; thanks to the design of http, it is the client / web-browser which has to be smart enough to download the file in several pieces. I'm not aware of any regular web browsers which do this, but "download managers" / "download accelerators" is what you want. [DownThemAll](https://www.downthemall.org/) used to be a very good choice, but I'm not sure if that is still the case... As for command line downloaders, aria2c will get the job done.

I'll take a look and see what I can find, been a while since I looked into this :>

**Reply by Anchovy8629** [reply]
Thanks, i tried hosting again but this time the friend downloading had the same ISP as me, my upload speed went at full bandwidth.. and hes not even close to me like the others.. i guess there is no work around..

**Reply by 9001** [reply]
That makes sense -- ISPs have a good grip of the traffic inside their own networks, so you're more likely to hit slowdowns once you cross that boundary.

But yes, there is no quick fixes for this one I'm afraid -- unless 10 MiB/s per download is enough, or unless your friends would be happy with installing a browser extension to boost the download speed. I can't really think of any alternatives to copyparty which would be a better choice in this case either, so let me know if you encounter any :-)

I had a quick look at browser extensions which could help, and this one seems promising: https://add0n.com/turbo-download-manager-v2.html -- with that, you can rightclick files and select to download them using Turbo Download Manager for a nice boost.

But just to mention it, there actually is a solid workaround which would give everyone good speeds right away -- it's a fair bit of work to set up, so probably more trouble than it's worth in your case, but here's the thing: You could rent a small VPS, and use that as a tunnel / reverse-proxy for your server at home. So people connect to the VPS, and the VPS forwards the traffic to you, and all the replies go back out through the VPS. This sounds silly, but it actually works great since the VPS providers make sure they have good routes everywhere :-) The demo server is such a setup; https://a.ocv.me/ is going to a vps in finland, which forwards all of the connections to the small server in my basement in norway. Costs me $5/month but it's worth it, since it's useful for lots of other stuff as well!

Thanks for trying copyparty btw :+1: :+1:

**Reply by 9001** [reply]
Actually wait, scratch that -- this will probably work!

[Cloudflare](https://www.cloudflare.com/) is a free service which basically does the tunnel thing for you, just way easier and without any maintenance. If you create an account there, you can add a DNS entry pointing to your home IP and enable proxying (yellow cloud). Then all the traffic will go through cloudflare's servers, which should be blazing fast for everyone :>

![Screenshot_2023-05-31_23-15-28-or8](https://github.com/9001/copyparty/assets/241032/445b7f11-4b3f-41aa-8560-41fafa39c4a2)

Just note that you will need to run your copyparty on [one of the following ports](https://developers.cloudflare.com/fundamentals/get-started/reference/network-ports/) if you use HTTP: 80, 8080, 8880, 2052, 2082, 2086, 2095

or one of these ports if you use HTTPS: 443, 2053, 2083, 2087, 2096, 8443

**Reply by Anchovy8629** [reply]
Thanks, i think i'm too stupid to understand this.. this works for websites, right? so i need a domain and i can't use something like no-ip right?

**Reply by 9001** [reply]
Ah yeah, that's right -- cloudflare does require you to own a "real" domain... So you'd have to buy one, and set up the domain to use cloudflare as its nameserver. Cloudflare should have some great guides on how to get through the entire process once you've bought a domain from somewhere :>

If you decide to buy a domain and give it a go, I think I'd recommend buying the domain from porkbun -- just make sure you get a domain which allows you to enable "whois privacy" (and remember to actually enable it!), and use your real name and address when registering.

**Reply by Anchovy8629** [reply]
Maybe i can use [this](https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/do-more-with-tunnels/trycloudflare/)? I can use my IP on this and it spats out a link to use, but i still don't know if i can setup the proxy part. I'm totally going blind on this and i know literally nothing about what i'm doing btw. Thanks for the help.

EDIT: changed the url, i'm using "Quick tunnels" with cloudflared, if can enable the proxy with this it should work i think?

**Reply by 9001** [reply]
Neat! Hadn't heard about that, looks like it should work yeah :>

**Reply by Anchovy8629** [reply]
I managed to use the cloudflared "tunnel" but i don't know if that by itself means it has proxy, the speeds are the same anyway.. but at least i don't have to open ports on the router..

**Reply by 9001** [reply]
Huh, that's a surprise -- tried the quick tunnel just now to verify, and yeah you did get proxied / tunneled through cloudflare's network, so exactly the same thing as if you bought a domain like I initially suggested. Really unexpected that it didn't improve the speed...

We can hope that the slow route is a temporary thing; maybe the ISP is working on it -- could try again in a few days/weeks and see if anything changed. But right now I don't think there's much more we can do on our end.

at least we learned some stuff :+1:

**Reply by Anchovy8629** [reply]
I'm not so sure that using quick tunneling means you get proxy too. I found [this](https://noted.lol/say-goodbye-to-reverse-proxy-and-hello-to-cloudflare-tunnels/) that goes more in depth with some configuration, but i still have to try.

**Reply by 9001** [reply]
A quick way to check if you got proxied is by adding these arguments when launching copyparty:
```
--ihead cf-ipcountry --ihead cf-visitor
```
the cloudflare proxy will insert some extra HTTP headers into the requests, so now we'll be able to see some of those in the log --
![Screenshot_2023-06-01_17-32-47-or8](https://github.com/9001/copyparty/assets/241032/57ef00f2-165d-49c3-a749-d5febb92ffd5)
so for each request that copyparty receives, it will then also display one or more lines with `[H] cf-...` if that request got proxied through cloudflare :+1: 

and if you visit using the no-ip link it'll show just the request and no cf-lines, since it didn't get proxied :>

EDIT: oh and if your friends' ISP is doing some shady stuff, there might even be a speed difference if you use either http:// or https:// with the cloudflare link btw -- but then there's even more reason to use https...
